src/java_bytecode/java_bytecode_typecheck_expr.cpp

```diff
@@ -164,18 +167,36 @@ void java_bytecode_typecheckt::typecheck_expr_java_string_literal(exprt &expr)
     // These are basically const global data:
     array_symbol.is_static_lifetime=true;
     array_symbol.is_state_var=true;
-    auto literal_array=utf16_to_array(
-      utf8_to_utf16_little_endian(id2string(value)));
-    array_symbol.value=literal_array;
+    array_symbol.value=
+      utf16_to_array(utf8_to_utf16_little_endian(id2string(value)));
+    array_symbol.type=array_symbol.value.type();
 
     if(symbol_table.add(array_symbol))
       throw "failed to add constarray symbol to symbol table";
 
-    literal_init.copy_to_operands(
-      from_integer(literal_array.operands().size(),
-                   jls_struct.components()[1].type()));
-    literal_init.copy_to_operands(
-      address_of_exprt(array_symbol.symbol_expr()));
+    const symbol_exprt array_expr=array_symbol.symbol_expr();
+    const address_of_exprt first_index(
+      index_exprt(array_expr, from_integer(0, java_int_type())));
+    literal_init.copy_to_operands(first_index);
+
+    // Associate array with pointer
+    symbolt return_symbol;
+    return_symbol.name="return_value_"+escaped_symbol_name;
+    return_symbol.base_name=escaped_symbol_name;
+    return_symbol.pretty_name="return_value";
+    return_symbol.mode=ID_java;
+    return_symbol.is_type=false;
+    return_symbol.is_lvalue=true;
+    return_symbol.is_static_lifetime=true;
+    return_symbol.is_state_var=true;
+    return_symbol.value=make_function_application(
```

Comment:
This use of `make_function_application isn't very readable, because the name `associate_array_to_pointer` suggests we ought to be passing an array and a pointer. How about wrapping the `make_function_application` call into something like `get_string_object_from_char_array`?

Reply:
> because the name `associate_array_to_pointer` suggests we ought to be passing an array and a pointer
I'm not sure what you mean here because we are passing an array `array_symbol.value` and a pointer `first_index` to this function.

Reply:
Aha, in that case rename `first_index` since I thought that was an index (specifically the first one ;))
Assuming that is &a[0] I'd just call that array_pointer.